yopt/src/test/scala/pravda/yopt/CommandLineTest.scala

```diff
@@ -20,15 +22,91 @@ object CommandLineTest extends TestSuite {
       groupParser("-abab")
     }
 
-    "int parser should parse 1" - {
-      intParser("--foo", "1")
-      intParser("-f", "1")
-      intParser("--foo", "0x01")
-      intParser("-f", "0x1")
-      intParserFail("--foo")
-      intParserFail("--foo", "bar")
+    "int parser should parse" - {
+      primTypesParser[Int](1, "--integer", "1")(_.intValue)
+      primTypesParser[Int](1, "--integer", "0x01")(_.intValue)
+      primTypesParser[Int](1, "--integer", "0x1")(_.intValue)
+      primTypesParser[Int](255, "--integer", "0xFF")(_.intValue)
+    }
+    
+    "int parser should fail" - {
+      primTypesParserFail("--integer", "FF")
+      primTypesParserFail("--integer")
+      primTypesParserFail("--integer", "bar")
+      primTypesParserFail("--integer", "-90")
+    }
+    
+    "long parser should parse" - {
+      primTypesParser[Long](999L, "--long", "999")(_.longValue)
+      primTypesParser[Long](15L, "--long", "0xF")(_.longValue)
+    }
+    
+    "bool parser should parse" - {
+      primTypesParser[Boolean](true, "--boolean", "true")(_.boolValue)
+      primTypesParser[Boolean](false, "--boolean", "false")(_.boolValue)
+      primTypesParser[Boolean](true, "--boolean", "yes")(_.boolValue)
+      primTypesParser[Boolean](false, "--boolean", "no")(_.boolValue)
+      primTypesParser[Boolean](false, "--boolean", "n")(_.boolValue)
+      primTypesParser[Boolean](true, "--boolean", "y")(_.boolValue)
+    }
+    
+    "bigdecimal parser should parse" - {
+      primTypesParser[BigDecimal](1.0, "--bigdecimal", "1.0")(_.bigDecimalValue)
+      primTypesParser[BigDecimal](200L, "--bigdecimal", "200")(_.bigDecimalValue)
+      primTypesParser[BigDecimal](100, "--bigdecimal", "100")(_.bigDecimalValue)
+    }
+
+    "double parser should parse" - {
+      primTypesParser[Double](1.0, "--double", "1.0")(_.doubleValue)
+      primTypesParser[Double](200L, "--double", "200")(_.doubleValue)
+      primTypesParser[Double](100, "--double", "100")(_.doubleValue)
+    }
+    
+    "uri parser should parse" - {
+      otherTypesParser[URI](new URI("http://localhost"), "--uri", "http://localhost")(_.uriValue)
+      otherTypesParser[URI](new URI("https://localhost"), "--uri", "https://localhost")(_.uriValue)
+      otherTypesParser[URI](new URI("http://localhost:8080/api/public/broadcast"), "--uri",
+        "http://localhost:8080/api/public/broadcast")(_.uriValue)
+    }
+  
+    // FIXME The standard URI class checks almost nothing. It even does not catch a simple typos
```

Comment:
Indeed! Because URI is a very vague format: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier

Reply:
We are not going to merge additional FIXMEs :) We have enough ours.

Reply:
I think the task of writing properly parser for URI can be extracted to a separated todo-task in that issue, that can be done by another PR. What do you think?

Reply:
`http:/localhost` is a valid URI, as I understand. So the question is do we really need URI parser?

Reply:
No, "http:/localhost" is not a valid URI, since here there is only one slash instead of obligatory two.

Valid URI should follow this grammar:
URI = scheme:[//authority]path[?query][#fragment]

Reply:
```
scheme = http
path = /localhost
```
:smirk:

Reply:
No, two forward slashes are obligatory. If "http" scheme is used then URI should comply the following syntax:
```
http[s]://<user>:<password>@<host>:<port>/<url-path>
```
Some parts can be omitted. See http://www.ietf.org/rfc/rfc1738.txt
Also here https://www.w3.org/Addressing/URL/5_BNF.html the BNF is described.

P.S. If ftp scheme is used and we omit "localhost" then we should use 3 forward slashes like this:

ftp:///C:/pagefile.sys

Reply:
:sweat: My opinion is we shoudn't add such complicated logic for the generic library that parse cmd args. Nevertheless, we should provide some mechanism for users to create and validate user-specific sophisticated data. And we have implicit instances for that purpose. So I suggest to exclude any non-trivial parsers from standard capabilities of the library and focus more on extensibility and "syntactic sugar".